### PR TITLE
[SPARK-48583][SQL][TESTS] Replace deprecated classes and methods of `commons-io` called in Spark

### PR DESCRIPTION
**Pull request by wayneguow**
### What changes were proposed in this pull request?

This pr replaces deprecated classes and methods of `commons-io`  called in Spark：

- `writeStringToFile(final File file, final String data)` -> `writeStringToFile(final File file, final String data, final Charset charset)`
- `CountingInputStream` -> `BoundedInputStream`

### Why are the changes needed?

Clean up deprecated API usage.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Passed related test cases in `UDFXPathUtilSuite` and `XmlSuite`.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by LuciferYang** [comment]
Is this the only case related to commons-io?

**Comment by wayneguow** [comment]
> Is this the only case related to commons-io?

For `FileUtils#writeStringToFile`, this is the only one left. Do we need to fix all other deprecated methods of `commons-io` called in Spark?

**Comment by LuciferYang** [comment]
Yes, if there are other cases related to commons-io, let's fix them together all in one

**Comment by wayneguow** [comment]
> @wayneguow seems there are other cases related to commons-io in `org.apache.spark.sql.execution.datasources.xml.XmlRecordReader`

@LuciferYang Thank you for pointing that out. I'll fix this and other left in a batch.

**Comment by wayneguow** [comment]
@LuciferYang I made a double check, the following `commons-io `classes are used in Spark:

- org.apache.commons.io.filefilter.TrueFileFilter
- org.apache.commons.io.FilenameUtils
- org.apache.commons.io.FileUtils
- org.apache.commons.io.input.CountingInputStream
- org.apache.commons.io.IOUtils
- org.apache.commons.io.output.ByteArrayOutputStream
- org.apache.commons.io.output.TeeOutputStream

In the build log, only `FileUtils` and `CountingInputStream` appear with deprecated prompts.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/xml/XmlInputFormat.scala [hunk 1]
```diff
@@ -117,7 +117,10 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
       }
     } else {
       fsin.seek(start)
-      countingIn = new CountingInputStream(fsin)
+      countingIn = BoundedInputStream.builder()
+        .setInputStream(fsin)
+        .setMaxCount(fileSplit.getLength)
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/xml/XmlInputFormat.scala:122]: Is this really necessary in the current scenario? Perhaps the `unbound` one is enough?

Comment (wayneguow) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/xml/XmlInputFormat.scala]: Sounds good, changed to `unbound `.

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @wayneguow

**Pull request closed without merging**